### PR TITLE
Add a gesture to open up a context menu; fix quitting

### PR DESCRIPTION
**Pull request by Manishearth**
This adds a gesture (currently: "palm facing upwards") to open up a context menu.

servo side: https://github.com/servo/servo/pull/26051

r? @jdm

### CODE REVIEW

**Comment by Manishearth** [comment]
A better version of this gesture would include head pose. I think the original suggested gesture was "looking at the palm of your hand", perhaps with the hand pointing upwards.

File: webxr/openxr/input.rs [hunk 1]
```diff
@@ -211,11 +218,53 @@ impl OpenXRInput {
         session: &Session<D3D11>,
         frame_state: &FrameState,
         base_space: &Space,
-    ) -> (InputFrame, Option<SelectEvent>, Option<SelectEvent>) {
+    ) -> (
+        InputFrame,
+        /* clicked */ Option<SelectEvent>,
+        /* squeezed */ Option<SelectEvent>,
+        /* menu_selected */ bool,
```
Comment (jdm) [inline webxr/openxr/input.rs]: Returning a real structure with named fields might be useful at this point.

File: webxr/openxr/input.rs [hunk 2]
```diff
@@ -211,11 +218,53 @@ impl OpenXRInput {
...
+    ) {
+        use euclid::Vector3D;
         let target_ray_origin = pose_for(&self.action_aim_space, frame_state, base_space);
 
         let grip_origin = pose_for(&self.action_grip_space, frame_state, base_space);
 
+        let mut menu_selected = false;
+        // Check if the palm is facing up. This is our "menu" gesture.
+        if let Some(grip_origin) = grip_origin {
+            // The X axis of the grip is perpendicular to the palm, however it's
```
Comment (jdm) [inline webxr/openxr/input.rs]: Nit: its

File: webxr/openxr/input.rs [hunk 3]
```diff
@@ -211,11 +218,53 @@ impl OpenXRInput {
...
+            // direction is the opposite for each hand
+            //
+            // We obtain a unit vector poking out of the palm
+            let x_dir = if let Handedness::Left = self.handedness {
+                1.0
+            } else {
+                -1.0
+            };
+
+            // Rotate it by the grip to btain the desired vector
```
Comment (jdm) [inline webxr/openxr/input.rs]: Nit: obtain

File: webxr/openxr/input.rs [hunk 4]
```diff
@@ -211,11 +218,53 @@ impl OpenXRInput {
...
+                .transform_vector3d(Vector3D::new(x_dir, 0.0, 0.0));
+
+            // Dot product it with the "up" vector to see if it's pointing up
+            let angle = grip_x.dot(Vector3D::new(0.0, 1.0, 0.0));
+
+            // If the angle is close enough to 0, its cosine will be
+            // close to 1
+            if angle > 0.9 {
+                self.menu_gesture_sustain += 1;
+                if self.menu_gesture_sustain > 60 {
```
Comment (jdm) [inline webxr/openxr/input.rs]: Should we make this a named constant?

File: webxr/openxr/mod.rs [hunk 5]
```diff
@@ -65,19 +65,58 @@ pub trait SurfaceProviderRegistration: Send {
     fn clone(&self) -> Box<dyn SurfaceProviderRegistration>;
 }
 
+///
```
Comment (jdm) [inline webxr/openxr/mod.rs]: /// Provides a way to interact with context menus.

**Pull request merged**